flink-runtime/src/test/java/org/apache/flink/runtime/codegeneration/SortingTest.java

```diff
@@ -0,0 +1,617 @@
+	
+	private static final long SEED2 = 97652436586326573L;
+
+	private static final int KEY_MAX = Integer.MAX_VALUE;
+
+	private static final int VALUE_LENGTH = 118;
+
+	private static final int MEMORY_SIZE = 1024 * 1024 * 64;
+	
+	private static final int MEMORY_PAGE_SIZE = 32 * 1024; 
+
+	private MemoryManager memoryManager;
+	private SorterFactory sorterFactory;
+
+
+	@Before
+	public void beforeTest() throws IOException {
+		this.memoryManager = new MemoryManager(MEMORY_SIZE, 1, MEMORY_PAGE_SIZE, MemoryType.HEAP, true);
+		this.sorterFactory = SorterFactory.getInstance();
+	}
+
+	@After
+	public void afterTest() {
+		if (!this.memoryManager.verifyEmpty()) {
+			Assert.fail("Memory Leak: Some memory has not been returned to the memory manager.");
+		}
+		
+		if (this.memoryManager != null) {
+			this.memoryManager.shutdown();
+			this.memoryManager = null;
+		}
+	}
+
+	private InMemorySorter<Tuple2<Integer, String>> newSortBuffer(List<MemorySegment> memory) throws Exception {
+		return this.sorterFactory.createSorter(TestData.getIntStringTupleSerializer(), TestData.getIntStringTupleComparator(), memory);
+	}
+
+	@Test
+	public void testSortIntKeys() throws Exception {
+		final int NUM_RECORDS = 559273;
+
+		final int numSegments = MEMORY_SIZE / MEMORY_PAGE_SIZE;
+		final List<MemorySegment> memory = this.memoryManager.allocatePages(new DummyInvokable(), numSegments);
+
+		InMemorySorter<Tuple2<Integer, String>> sorter = newSortBuffer(memory);
+		TestData.TupleGenerator generator = new TestData.TupleGenerator(SEED, KEY_MAX, VALUE_LENGTH, KeyMode.RANDOM,
+			ValueMode.RANDOM_LENGTH);
+
+		// write the records
+		Tuple2<Integer, String> record = new Tuple2<>();
+		int num = 0;
+		do {
+			generator.next(record);
+			num++;
+		}
+		while (sorter.write(record) && num < NUM_RECORDS);
+
+		QuickSort qs = new QuickSort();
+		qs.sort(sorter);
+
+		MutableObjectIterator<Tuple2<Integer, String>> iter = sorter.getIterator();
+		Tuple2<Integer, String> readTarget = new Tuple2<>();
+
+		iter.next(readTarget);
+		int last = readTarget.f0;
+
+		while ((readTarget = iter.next(readTarget)) != null) {
+			int current = readTarget.f0;
+
+			final int cmp = last - current;
+			if (cmp > 0) {
+				Assert.fail("Next key is not larger or equal to previous key.");
+			}
+
+			last = current;
+		}
+
+		// release the memory occupied by the buffers
+		sorter.dispose();
+		this.memoryManager.release(memory);
+	}
+
+	@Test
+	public void testSortStringKeys() throws Exception {
+
+		@SuppressWarnings("unchecked")
+		List<MemorySegment> memory = createMemory();
+		TypeComparator<Tuple2<Integer, String>> accessors = TestData.getIntStringTupleTypeInfo().createComparator(new int[]{1}, new boolean[]{true}, 0, null);
+		InMemorySorter<Tuple2<Integer, String>> sorter = createSorter(TestData.getIntStringTupleSerializer(), accessors, memory);
+
+		TestData.TupleGenerator generator = new TestData.TupleGenerator(SEED, KEY_MAX, 5, KeyMode.RANDOM,
+			ValueMode.FIX_LENGTH);
```

Comment:
Is this testing the situation when the string is longer that the normalized key? (i.e., when the `else` branches are hit in the not fully determined case)